url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/ui/KlangProgressWindow.java

```diff
@@ -0,0 +1,599 @@
+	 * @param newChunkName the name for the new chunk
+	 * @param position the position of the new chunk (1 = first child)
+	 * @param fromFile this file will be read in its entirety and inserted into the chunk as data;
+	 * 		will be null if a new empty chunk should be added instead
+	 */
+	public void addChunk(KlangFile theFile, Chunk parentChunk, String newChunkName, int position, File fromFile) {
+		finished = false;
+		errMsg = "";
+		ProgressTask lTask = new ProgressTask(theTimer);
+        theTimer.schedule(lTask, 0, 200);
+        this.prgProg.setVisible(true);
+        this.lblProg.setVisible(true);
+        KlangProgressWindow.currStatus = "Adding Chunk...";
+        KlangProgressWindow.progVal = 0;
+        final KlangFile THEFILE = theFile;
+        final Chunk PARENTCHUNK = parentChunk;
+        final String NEWCHUNKNAME = newChunkName;
+        final Integer POSITION = position;
+        final File FROMFILE = fromFile;
+        Runnable q = new Runnable() {
+            public void run() {
+            	RandomAccessFile raf = null;
+		        try {
+		        	raf = new RandomAccessFile(THEFILE.getFile(), "rw");
+		        	if (PARENTCHUNK != null) {
+		        		if (FROMFILE != null) {
+		        			((EditableContainerChunk)PARENTCHUNK).addChunkFromFile(FROMFILE, NEWCHUNKNAME, POSITION, raf);
+		        		} else {
+		        			((EditableContainerChunk)PARENTCHUNK).addChunkFromArray(new byte[0], NEWCHUNKNAME, POSITION, raf);
+		        		}
+		        	} else {
+		        		EditableFileBase efb = (EditableFileBase)THEFILE;
+		        		if (FROMFILE != null) {
+		        			efb.addChunkFromFile(FROMFILE, NEWCHUNKNAME, POSITION, raf);
+		        		} else {
+		        			efb.addChunkFromArray(new byte[0], NEWCHUNKNAME, POSITION, raf);
+		        		}
+		        	}
+		        	raf.close();
+		        	finished = true;
+				} catch (Exception err) {
+					if (raf != null) {
+						try {
+							raf.close();
+						} catch (Exception errr) {}
+					}
+					err.printStackTrace();
+					finished = true;
+					errMsg = err.getMessage();
+					return;
+				}
+            }
+        };
+        Thread t = new Thread(q);
+        t.start();
+        this.setVisible(true);
+	}
+	
+	/**
+	 * Delete the specified chunk.
+	 * @param chunk the chunk to delete
+	 */
+	public void deleteChunk(Chunk chunk) {
+		finished = false;
+		errMsg = "";
+		ProgressTask lTask = new ProgressTask(theTimer);
+        theTimer.schedule(lTask, 0, 200);
+        this.prgProg.setVisible(true);
+        this.lblProg.setVisible(true);
+        KlangProgressWindow.currStatus = "Deleting Chunk...";
+        KlangProgressWindow.progVal = 0;
+        final Chunk CHUNK = chunk;
+        Runnable q = new Runnable() {
+            public void run() {
+            	RandomAccessFile raf = null;
+		        try {
+		        	if (!(CHUNK.chunkCanBeDeleted())) {
+		        		throw new IOException(KlangConstants.ERROR_CHUNK_ISNT_EDITABLE);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [IOException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [281].  Consider adding a catch clause of this specific exception in the outer try statement.